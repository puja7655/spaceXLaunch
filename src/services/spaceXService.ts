export const getLaunches = async (offset?: number) => {
    try {
      const response = await fetch(`https://api.spacexdata.com/v4/launches?limit=10&offset=${offset}`);
      if (!response.ok) {
        throw new Error('Failed to fetch launches');
      }
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error fetching launches:', error);
      throw error;
    }
  };
  