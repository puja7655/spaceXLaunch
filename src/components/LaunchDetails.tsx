import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { getLaunches } from '../services/spaceXService';

interface LaunchDetailsProps {
    launchId?: string;
}

const LaunchDetails: React.FC<LaunchDetailsProps> = ({ launchId }) => {
    const [launch, setLaunch] = useState<any | null>(null);
    const { launchId: paramLaunchId } = useParams<{ launchId: string }>();

    useEffect(() => {
        const fetchLaunch = async () => {
            try {
                const data = await getLaunches();
                const foundLaunch = data.find((launch: any) => launch.id === (launchId || paramLaunchId));
                setLaunch(foundLaunch);
            } catch (error) {
                console.error('Error fetching launch details:', error);
            }
        };

        fetchLaunch();
    }, [launchId, paramLaunchId]);

    if (!launch) {
        return (
            <div className="container mt-4">
                <p>Loading...</p>
            </div>
        );
    }

    return (
        <div className="container mt-4">
            <h2>{launch.name}</h2>
            <div className="row">
                <div className="col-md-6">
                    <img src={launch.links.patch.small} alt={launch.name} className="img-fluid" />
                </div>
                <div className="col-md-6">
                    <p><h4 className="list-group-item-heading">Details</h4></p>
                    <p>{launch.details}</p>
                    <p><h4 className="list-group-item-heading">Status</h4></p>
                    <p>{launch.success ? 'Successful' : 'Failed'}</p>
                    <p><h4 className="list-group-item-heading">Rocket</h4></p>
                    <p>{launch.rocket}</p>
                </div>
            </div>
        </div>
    );
};

export default LaunchDetails;
