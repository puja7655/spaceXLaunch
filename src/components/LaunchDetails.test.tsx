import { render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import LaunchDetails from './LaunchDetails';
import * as spaceXService from '../services/spaceXService';

const mockedUseParams = jest.fn();
jest.mock('react-router-dom', () => ({
    useParams: () => mockedUseParams(),
}));

jest.mock('../services/spaceXService', () => ({
    ...jest.requireActual('../services/spaceXService'),
    getLaunches: jest.fn(),
}));

describe('LaunchDetails component', () => {
    it('renders loading message when launch is not loaded', async () => {
        mockedUseParams.mockReturnValue({ launchId: 'exampleLaunchId' });

        (spaceXService.getLaunches as jest.Mock).mockResolvedValue([]);

        const { getByText } = render(<LaunchDetails />);

        expect(getByText('Loading...')).toBeInTheDocument();
        await waitFor(() => { });
    });

    it('renders launch details when launch is loaded', async () => {
        mockedUseParams.mockReturnValue({ launchId: 'exampleLaunchId' });

        (spaceXService.getLaunches as jest.Mock).mockResolvedValue([]);

        const { container } = render(<LaunchDetails />)
        container.querySelector('.list-group-item-heading')
        await waitFor(() => { });
        expect(container.querySelector('.list-group-item-heading')).toBeDefined();
        expect(container.firstChild).toMatchSnapshot()
    });
});
