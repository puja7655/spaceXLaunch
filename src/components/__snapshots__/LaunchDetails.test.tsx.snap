// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchDetails component renders launch details when launch is loaded 1`] = `
<div
  class="container mt-4"
>
  <p>
    Loading...
  </p>
</div>
`;
